definitions/npm/async_v1.x.x/flow_>=v0.28.x/async_v1.x.x.js

```diff
@@ -0,0 +1,77 @@
+// @flow
+
+declare module 'async' {
+  declare type Collection<TItem> = Iterable<TItem> | { [key: string]: TItem };
+
+  declare type IterateeCallback<TError, TResult> = (err?: TError, result?: TResult) => void;
+  declare type Callback<TError, TResult> = (err?: TError, result?: TResult) => mixed;
+
+  // Collections
+  declare function each<TItem, TError>(
+    collection: Collection<TItem>,
+    iteratee: (item: TItem, callback: IterateeCallback<TError, void>) => mixed,
+    callback?: Callback<TError, void>
+  ): void;
+  declare function forEachOf<TItem, TError>(
+    collection: Collection<TItem>,
+    iteratee: (value: TItem, key: string | number, callback: IterateeCallback<TError, void>) => mixed,
```

Comment:
What I meant by "overload this function using each of the two members of the union and writing out their specific" was something like

```js
declare function forEachOf<TValue, TError>(
  collection: Array<TValue>,
  iteratee: (value: TValue, key: number, ...) => mixed,
  ...
);

declare function forEachOf<TValue, TError>(
  collection: {[key: string]: TValue},
  iteratee: (value: TValue, key: string, ...) => mixed,
  ...
);

declare function forEachOf<TValue, TError>(
  collection: Iterator<TValue>,
  iteratee: (value: TValue, key: string | number, ...) => mixed,
  ...
);
```

I'm relatively sure Flow (v0.28) won't know how to choose these either, unfortunately (cc @avikchaudhuri here as well).

In the meantime, I think what you have here already is probably the best amount of precision you can get away with.